File: app/Libraries/Morphology/AffixInfoProvider.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MorfParsingLibrary
{
    class AffixInfoProvider
    {
        //ФУНКЦИИ ПРОВЕРКИ АФФИКСОВ
        //перечисления для указания типа поиска и типа восстановления
        public enum SearchMode
        {
            OnlyGlagol, NotGlagol, Any, Prichastie, DeePrichastie, Prilagatelnoe, NotNarechie, NotPril,
            NotNarechieGlagol, SuschGlagol, Chislitelnoe, deenoun, AbsolutelyAny
        }
        public enum RecoveryMode
        {
            With, Without, Both
        }
        //Метод присоединения нового аффикса к юниту аффиксов
        public static string AddAffixToUnit(string aff, string AffUnit)
        {
            if (!(AffUnit.Length == 0))
            {
                if (AffUnit[AffUnit.Length - 1] == '|')
                {
                    AffUnit = AffUnit.Substring(0, AffUnit.Length - 1);
                }
            }

            if (aff.StartsWith("|"))
            {
                return aff + AffUnit;
            }
            return ("|" + aff + AffUnit);
        }
        //Метод проверки "подходимости" аффикса к аффиксу
        public static bool AffixesSuitable(string aff, string AffUnit)
        {
            //типы
            string afftype, unittype;
            bool decision = false;
            //Уровень аффикса
            int afflevel = LevelOfSingleAffix(aff, out afftype);
            //Уровень юнита
            int unitlevel = LevelOfAffUnit(AffUnit, out unittype);
            

            //Trace.WriteLine(aff + ":" + "уровень:" + afflevel + "тип:" + afftype);
            //Trace.WriteLine(AffUnit + ":" + "уровень:" + unitlevel + "тип:" + unittype);
            if (afflevel < unitlevel)
            {
                if ((afftype == unittype) || (afftype == Constants.ANY) || (unittype == Constants.ANY))
                {
                    decision = true;
                }
            }
            //Trace.WriteLine(decision);

            return decision;

        }
        //Метод определения "уровня", типа единичного аффикса
        public static int LevelOfSingleAffix(string AffixForCheck, out string singleafftype)
        {
            singleafftype = Constants.ANY;
            int level = 888;
            
            string[] affmas = AffixForCheck.Split(new char[] { '|' }, StringSplitOptions.RemoveEmptyEntries);
            switch (AffixForCheck)
            {
                case "": level = 100;
                    break;
                case "|": level = 90;
                    break;
                default:

                    for (int i = MorfParser.Affixes.Length - 1; i >= 0; i--)
                    {
                        string[] affarr = MorfParser.Affixes[i].Split(new char[] { ',', ' ' }, StringSplitOptions.RemoveEmptyEntries);
                        for (int j = 0; j < affarr.Length; j++)
                        {
                            if (affarr[j] == affmas[0])
                            {
                                singleafftype = affarr[affarr.Length - 1];
                                return i;                               
                            }

                        }
                    }
                    
                    break;
            }
            return level;
        }
        //Метод определения "уровня", типа юнита аффиксов
        public static int LevelOfAffUnit(string AffUnitForCheck, out string affunittype)
        {
            affunittype = Constants.ANY;

            if (AffUnitForCheck == "") return 100;
            if (AffUnitForCheck == "|") return 90;

            string[] demo = { Constants.ONLYGLAGOL, Constants.NOTGLAGOL };
            
            string singleafftype;
            string[] affmas = AffUnitForCheck.Split(new char[] { '|' }, StringSplitOptions.RemoveEmptyEntries);

            int level = LevelOfSingleAffix(affmas[0], out singleafftype);
            
            for (int i = 0; i < affmas.Length; i++)
            {
                LevelOfSingleAffix(affmas[i], out singleafftype);
                if (demo.Contains(singleafftype))
                {
                    affunittype = singleafftype;
                    return level;
                }
            }
            return level;

            
        }
        //Метод определения типа поиска в словаре
        public  static SearchMode TypeOfSearch(string AffUnit)
        {
            SearchMode mode = SearchMode.Any;
            string afftype;
            int level = LevelOfSingleAffix(AffUnit, out afftype);

            if (level >= 50) return mode;


            string[] aff = AffUnit.Split(new char[] { '|' }, StringSplitOptions.RemoveEmptyEntries);
            for (int i = 0; i < aff.Length; i++)
            {
                LevelOfSingleAffix(aff[i], out afftype);
                switch (aff[i])
                {
                    case "ни":
                    case "акан":
                    case "екен":
                    case "аканни":
                    case "екенни":
                    case "малли":
                    case "мелли":
                    case "малла":
                    case "мелле":
                    case "манни":
                    case "менни":
                    case "нӑ":
                    case "нӗ":
                    case "ас":
                    case "ес":
                    case "асси":
                    case "есси":
                    case "асшӑн":
                    case "есшӗн":
                    //case "ен":
                    //case "ан":
                    //case "мен":
                    //case "ман":
                    case "нӑҫем":
                    case "нӗҫем":
                        mode = SearchMode.Prichastie;
                        goto metka;
                    case "са":
                    case "се":
                    case "сан":
                    case "сассӑн":
                    case "сессӗн":
                    case "масӑр":
                    case "месӗр":
                        mode = SearchMode.DeePrichastie;
                        goto metka;
                    case "а":
                    case "е":
                    case "сен":
                        mode = SearchMode.deenoun;
                        goto metka;
                    case "лӑ":
                    case "лӗ":
                    case "ллӑ":
                    case "ллӗ":
                    case "ли":
                    case "лли":
                        mode = SearchMode.Prilagatelnoe;
                        goto metka;
                    case "мӑш":
                    case "мӗш":
                        mode = SearchMode.Chislitelnoe;
                        goto metka;
                    case "на":
                    case "не":
                        mode = SearchMode.NotNarechieGlagol;
                        goto metka;
                    //case "ах":
                    //case "ех":
                    //case "х":
                    //    mode = SearchMode.AbsolutelyAny;
                    //    goto metka;
                    default:
                        break;
                }

                switch (afftype)
                {
                    case Constants.ONLYGLAGOL:
                    case Constants.NOTGLAGOL:
                        mode = (SearchMode)Enum.Parse(typeof(SearchMode), afftype);
                        break;
                    default:
                        break;
                }
            }
        metka:
            return mode;
        }
        //Метод определения типа восстановления корня
        public static RecoveryMode TypeOfRecovery(string AffUnit)
        {
            RecoveryMode mode = RecoveryMode.Both;
            //выделяем первый аффикс
            string aff = AffUnit.Split(new char[] { '|' }, StringSplitOptions.RemoveEmptyEntries)[0];

            switch (aff)
            {
                case "и":
                case "у":
                    mode = RecoveryMode.With;
                    break;
                case "м":
                case "а":
                case "е":
                case "ти":
                case "ри":
                case "рен":
                case "аттӑм":
                case "еттӗм":
                case "аттӑн":
                case "еттӗн":
                case "ттӑм":
                case "ттӗм":
                case "ттӑн":
                case "ттӗн":
                case "атчӗҫ":
                case "етчӗҫ":
                case "тчӗҫ":
                case "сассӑн":
                case "тӑп":
                case "тӗп":
                case "тӑн":
                case "тӗн":
                case "атӑп":
                case "етӗп":
                case "атӑн":
                case "етӗн":
                case "тпӑр":
                case "тпӗр":
                case "атпӑр":
                case "етпӗр":
                //case "аҫ":
                //case "еҫ":
                case "атт":
                case "етт":
                case "тан":
                case "тен":
                case "та":
                case "те":
                case "ла":
                case "ле":
                case "сӑр":
                case "сӗр":
                case "сӑмӑр":
                case "сӗмӗр":
                case "ам":
                case "ем":
                case "ӑм":
                case "ӗм":
                case "тӑм":
                case "тӗм":
                case "на":
                case "не":
                case "ра":
                case "ре":
                case "сен":
                case "хи":
                case "ччен":
                case "чен":
                case "серен":
                case "сем":
                case "н":
                case "алла":
                case "елле":
                case "па":
                case "пе":
                case "сам":
                case "ран":
                case "сан":
                case "сене":
                case "ма":
                case "ме":
                case "масӑр":
                case "месӗр":
                //case "ат":мешает при корне "выля"
                case "ет":
                    mode = RecoveryMode.Without;
                    break;
            }
            return mode;
        }
        //определение искомой части речи
        public static string PatternPicker(string AffUnit)
        {
            SearchMode sm = TypeOfSearch(AffUnit);

            Trace.WriteLine("Тип поиска: [" + sm + "]");

            string pattern = "";
            switch (sm)
            {
                case SearchMode.OnlyGlagol:
                    pattern =  Constants.VERB + "|Same";
                    break;
                case SearchMode.NotGlagol:
                    pattern = Constants.NOUN +  "," + Constants.NUMERIC + "," + Constants.ADJECTIVE + "," + Constants.PRONOUN + "," + Constants.ADVERB + "|Same";
                    break;
                case SearchMode.Any:
                    pattern = Constants.NOUN + "," + Constants.NUMERIC + "," + Constants.ADJECTIVE + "," + Constants.PRONOUN + "," + Constants.VERB + "|Same";
                    break;
                //case SearchMode.AbsolutelyAny:
                //    pattern = Constants.NOUN + "," + Constants.NUMERIC + "," + Constants.ADJECTIVE + "," + Constants.PRONOUN + "," + Constants.VERB + Constants.ADVERB + "|Same";
                //    break;
                //case SearchMode.NotNarechie:
                //    pattern = "сущ-е, числ-е, прил-е, мест-е, глагол|Same";
                //    break;
                case SearchMode.NotNarechieGlagol:
                    pattern = Constants.NOUN +  "," + Constants.NUMERIC + "," + Constants.ADJECTIVE + "," + Constants.PRONOUN + "|Same";
                    break;
                case SearchMode.Prichastie:
                    pattern = Constants.VERB + "|" + Constants.PRICHASTIE;
                    break;
                case SearchMode.DeePrichastie:
                    pattern = Constants.VERB + "|" + Constants.DEEPRICHASTIE;
                    break;
                case SearchMode.Prilagatelnoe:
                    pattern = Constants.NOUN + "," + Constants.NUMERIC + "|" + Constants.ADJECTIVE;
                    break;
                case SearchMode.NotPril:
                    pattern = Constants.NOUN + "," + Constants.NUMERIC + "," + Constants.ADVERB + "," + Constants.PRONOUN + "," + Constants.VERB + "|Same";
                    break;
                case SearchMode.SuschGlagol:
                    pattern = Constants.NOUN + "," + Constants.VERB + "|Same";
                    break;
                case SearchMode.Chislitelnoe:
                    pattern = Constants.NUMERIC + "|Same";
                    break;
                case SearchMode.deenoun:
                    pattern = Constants.NOUN + "," + Constants.NUMERIC + "," + Constants.ADJECTIVE + "," + Constants.PRONOUN + "," + Constants.VERB + "," + Constants.ADVERB + "|" + Constants.DEENOUN;
                    break;
                default:
                    break;
            }
            return pattern;
        }
        //обрезка "полных" аффиксов
        public static string CutItOut(string AffUnit, string chast)
        {
            string[] affunit = AffUnit.Split(new char[] { '|' }, StringSplitOptions.RemoveEmptyEntries);
            string finAffUnit = "";
            string separ = "|";
            for (int i = 0; i < affunit.Length; i++)
            {
                switch (affunit[i])
                {
                    case "ҫӑм":
                    case "ҫӗм":
                        finAffUnit += affunit[i][0] + separ + affunit[i].Substring(1, 2) + separ; 
                        break;
                    case "ятна": 
                    case "атна": 
                    case "етне":
                        finAffUnit += affunit[i].Substring(0, 2) + separ + affunit[i].Substring(2, 2) + separ;
                        break;
                    case "масӑр":
                    case "месӗр":
                        finAffUnit += affunit[i].Substring(0, 2) + separ + affunit[i].Substring(2, 3) + separ;
                        break;
                    case "рӑм":
                    case "рӗм":
                    case "рӑн":
                    case "рӗн":
                    case "тӑм":
                    case "тӗм":
                    case "ман":
                    case "мен":
                    case "мӑп":
                    case "мӗп":
                    case "мап":
                    case "меп":
                    case "маҫ":
                    case "меҫ":
                    case "тӑп":
                    case "тӗп":
                    case "тӑн":
                    case "тӗн":
                    case "рӑр":
                    case "рӗр":
                        finAffUnit += affunit[i][0] + separ + affunit[i].Substring(1, 2) + separ;
                        break;
                    case "аттӑм":
                    case "еттӗм":
                    case "ӑттӑм":
                    case "ӗттӗм":
                        finAffUnit += affunit[i].Substring(0, 3) + separ + affunit[i].Substring(3, 2) + separ;
                        break;
                    case "ттӑм":
                    case "ттӗм":
                    case "ттӑн":
                    case "ттӗн":
                    case "атӑп":
                    case "етӗп":
                    case "атӑн":
                    case "етӗн":
                        finAffUnit += affunit[i].Substring(0, 2) + separ + affunit[i].Substring(2, 2) + separ;
                        break;
                    case "атчӗҫ":
                    case "етчӗҫ":
                        finAffUnit += affunit[i].Substring(0, 3) + separ + affunit[i].Substring(3, 1) + separ + affunit[i].Substring(4, 1) + separ;
                        break;
                    case "тчӗҫ":
                        finAffUnit += affunit[i].Substring(0, 2) + separ + affunit[i].Substring(2, 1) + separ + affunit[i].Substring(3, 1) + separ;
                        break;
                    case "аканни":
                    case "екенни":
                    case "яканни":
                        finAffUnit += affunit[i].Substring(0, 5) + separ + affunit[i].Substring(5, 1) + separ;
                        break;
                    case "манни":
                    case "менни":
                        finAffUnit += affunit[i][0] + separ + affunit[i].Substring(1, 3) + separ + affunit[i][4] + separ;
                        break;
                    case "асси":
                    case "есси":
                    case "ясси":
                        finAffUnit += affunit[i].Substring(0, 3) + separ + affunit[i][3] + separ;
                        break;
                    case "ни":
                        finAffUnit += "н" + separ + "и" + separ;
                        break;
                    case "малли":
                    case "мелли":
                        finAffUnit += affunit[i].Substring(0, 4) + separ + "и" + separ;
                        break;
                    case "нӑҫем":
                    case "нӗҫем":
                        finAffUnit += affunit[i].Substring(0, 2) + separ + affunit[i].Substring(2, 3) + separ;
                        break;
                    case "пӑр":
                    case "пӗр":
                        finAffUnit += affunit[i][0] + separ + affunit[i].Substring(1, 2) + separ;
                        break;
                    case "сене":
                        finAffUnit += "сем" + separ + "е" + separ;
                        break;
                    case "аҫҫӗ":
                    case "еҫҫӗ":
                        finAffUnit += affunit[i].Substring(0,2) + separ + "ҫ" + separ + "ӗ" + separ;
                        break;
                    case "мест":
                        finAffUnit += "мес" + separ + "т" + separ;
                        break;
                    case "масть":
                        finAffUnit += "мас" + separ + "ть" + separ;
                        break;
                    case "рӗ":
                        finAffUnit += affunit[i][0] + separ + "ӗ" + separ;
                        break;
                    case "рӗҫ":
                    case "чӗҫ":
                        finAffUnit += affunit[i][0] + separ + "ӗ" + separ + "ҫ" + separ;
                        break;
                    case "мӗ":
                        finAffUnit += "м" + separ + "ӗ" + separ;
                        break;
                    case "ӗҫ":
                        finAffUnit += "ӗ" + separ + "ҫ" + separ;
                        break;
                    case "атчӗ":
                    case "етчӗ":
                        finAffUnit += affunit[i][0] + "тч" + separ + "ӗ" + separ;
                        break;
                    case "чӗ":
                        if (chast == "глагол" && affunit[0] =="чӗ") finAffUnit += "ч" + separ + "ӗ" + separ;
                        else finAffUnit += affunit[i] + separ;
                        break;
                    default:
                        finAffUnit += affunit[i] + separ;
                        break;
                }
            }
            finAffUnit = finAffUnit.Remove(finAffUnit.Length - 1);
            return finAffUnit;
        }
    }
}

File: app/Services/MorfParsingLibrary/AffixInfoProvider.php
<?php
namespace App\Services\MorfParsingLibrary;


class AffixInfoProvider
{
    // Перечисления для указания типа поиска и типа восстановления
    const SearchMode = [
        'OnlyGlagol' => 0,
        'NotGlagol' => 1,
        'Any' => 2,
        'Prichastie' => 3,
        'DeePrichastie' => 4,
        'Prilagatelnoe' => 5,
        'NotNarechie' => 6,
        'NotPril' => 7,
        'NotNarechieGlagol' => 8,
        'SuschGlagol' => 9,
        'Chislitelnoe' => 10,
        'deenoun' => 11,
        'AbsolutelyAny' => 12,
    ];

    const RecoveryMode = [
        'With' => 0,
        'Without' => 1,
        'Both' => 2,
    ];

    // Метод присоединения нового аффикса к юниту аффиксов
    public static function addAffixToUnit($aff, $affUnit)
    {
        if (strlen($affUnit) > 0) {
            if ($affUnit[strlen($affUnit) - 1] == '|') {
                $affUnit = substr($affUnit, 0, -1);
            }
        }

        if (strpos($aff, '|') === 0) {
            return $aff . $affUnit;
        }

        return '|' . $aff . $affUnit;
    }

    // Метод проверки "подходимости" аффикса к аффиксу
    public static function affixesSuitable($aff, $affUnit)
    {
        // Типы
        $affType = '';
        $unitType = '';
        $decision = false;

        // Уровень аффикса
        $affLevel = self::levelOfSingleAffix($aff, $affType);

        // Уровень юнита
        $unitLevel = self::levelOfAffUnit($affUnit, $unitType);

        if ($affLevel < $unitLevel) {
            if ($affType === $unitType || $affType === Constants::ANY || $unitType === Constants::ANY) {
                $decision = true;
            }
        }

        return $decision;
    }

    
    // Метод определения "уровня", типа единичного аффикса
    public static function levelOfSingleAffix($affixForCheck, &$singleAffType)
    {
        $singleAffType = Constants::ANY;
        $level = 888;

        $affMas = explode('|', $affixForCheck);
        switch ($affixForCheck) {
            case '':
                $level = 100;
                break;
            case '|':
                $level = 90;
                break;
            default:
                // Используйте правильное имя статической переменной с правильным регистром
                for ($i = count(MorfParser::$Affixes) - 1; $i >= 0; $i--) {
                    $affArr = preg_split('/[,\s]+/', MorfParser::$Affixes[$i]);
                    if (in_array($affMas[0], $affArr)) {
                        $singleAffType = end($affArr);
                        return $i;
                    }
                }
                break;
        }

        return $level;
    }

    // Метод определения "уровня", типа юнита аффиксов
    public static function levelOfAffUnit($affUnitForCheck, &$affUnitType)
    {
        $affUnitType = Constants::ANY;

        if ($affUnitForCheck === '') {
            return 100;
        }
        if ($affUnitForCheck === '|') {
            return 90;
        }

        $demo = [Constants::ONLYGLAGOL, Constants::NOTGLAGOL];
        $singleAffType = '';
        $affMas = explode('|', $affUnitForCheck);
        $level = self::levelOfSingleAffix($affMas[0], $singleAffType);

        foreach ($affMas as $aff) {
            self::levelOfSingleAffix($aff, $singleAffType);
            if (in_array($singleAffType, $demo)) {
                $affUnitType = $singleAffType;
                return $level;
            }
        }

        return $level;
    }

    // Метод определения типа поиска в словаре
    public static function typeOfSearch($affUnit)
    {
        $mode = self::SearchMode['Any'];
        $affType = '';
        $level = self::levelOfSingleAffix($affUnit, $affType);

        if ($level >= 50) {
            return $mode;
        }

        $aff = explode('|', $affUnit);
        foreach ($aff as $a) {
            self::levelOfSingleAffix($a, $affType);
            switch ($a) {
                case 'ни':
                case 'акан':
                case 'екен':
                    $mode = self::SearchMode['Prichastie'];
                    break;
                case 'са':
                case 'се':
                    $mode = self::SearchMode['DeePrichastie'];
                    break;
                case 'а':
                case 'е':
                    $mode = self::SearchMode['deenoun'];
                    break;
                case 'лӑ':
                case 'лӗ':
                    $mode = self::SearchMode['Prilagatelnoe'];
                    break;
                default:
                    break;
            }

            switch ($affType) {
                case Constants::ONLYGLAGOL:
                case Constants::NOTGLAGOL:
                    $mode = array_search($affType, self::SearchMode);
                    break;
                default:
                    break;
            }
        }

        return $mode;
    }

    // Метод определения типа восстановления корня
    public static function typeOfRecovery($affUnit)
    {
        $mode = self::RecoveryMode['Both'];
        $aff = explode('|', $affUnit)[0];

        switch ($aff) {
            case 'и':
            case 'у':
                $mode = self::RecoveryMode['With'];
                break;
            case 'м':
            case 'а':
            case 'е':
                $mode = self::RecoveryMode['Without'];
                break;
        }

        return $mode;
    }

    // Метод определения искомой части речи
    public static function patternPicker($affUnit)
    {
        $sm = self::typeOfSearch($affUnit);

        $pattern = '';
        switch ($sm) {
            case self::SearchMode['OnlyGlagol']:
                $pattern = Constants::VERB . '|Same';
                break;
            case self::SearchMode['NotGlagol']:
                $pattern = Constants::NOUN . ',' . Constants::NUMERIC . ',' . Constants::ADJECTIVE . '|Same';
                break;
            case self::SearchMode['Any']:
                $pattern = Constants::NOUN . ',' . Constants::NUMERIC . ',' . Constants::VERB . '|Same';
                break;
        }

        return $pattern;
    }


    //обрезка "полных" аффиксов
    public static function CutItOut($AffUnit, $chast)
    {
        $affunit = array_filter(explode('|', $AffUnit)); // Разбиваем строку на массив и удаляем пустые элементы
        $finAffUnit = "";
        $separ = "|";

        foreach ($affunit as $aff) {
            switch ($aff) {
                case "ҫӑм":
                case "ҫӗм":
                    $finAffUnit .= $aff[0] . $separ . substr($aff, 1, 2) . $separ;
                    break;
                case "ятна":
                case "атна":
                case "етне":
                    $finAffUnit .= substr($aff, 0, 2) . $separ . substr($aff, 2, 2) . $separ;
                    break;
                case "масӑр":
                case "месӗр":
                    $finAffUnit .= substr($aff, 0, 2) . $separ . substr($aff, 2, 3) . $separ;
                    break;
                case "рӑм":
                case "рӗм":
                case "рӑн":
                case "рӗн":
                case "тӑм":
                case "тӗм":
                case "ман":
                case "мен":
                case "мӑп":
                case "мӗп":
                case "мап":
                case "меп":
                case "маҫ":
                case "меҫ":
                case "тӑп":
                case "тӗп":
                case "тӑн":
                case "тӗн":
                case "рӑр":
                case "рӗр":
                    $finAffUnit .= $aff[0] . $separ . substr($aff, 1, 2) . $separ;
                    break;
                case "аттӑм":
                case "еттӗм":
                case "ӑттӑм":
                case "ӗттӗм":
                    $finAffUnit .= substr($aff, 0, 3) . $separ . substr($aff, 3, 2) . $separ;
                    break;
                case "ттӑм":
                case "ттӗм":
                case "ттӑн":
                case "ттӗн":
                case "атӑп":
                case "етӗп":
                case "атӑн":
                case "етӗн":
                    $finAffUnit .= substr($aff, 0, 2) . $separ . substr($aff, 2, 2) . $separ;
                    break;
                case "атчӗҫ":
                case "етчӗҫ":
                    $finAffUnit .= substr($aff, 0, 3) . $separ . substr($aff, 3, 1) . $separ . substr($aff, 4, 1) . $separ;
                    break;
                case "тчӗҫ":
                    $finAffUnit .= substr($aff, 0, 2) . $separ . substr($aff, 2, 1) . $separ . substr($aff, 3, 1) . $separ;
                    break;
                case "аканни":
                case "екенни":
                case "яканни":
                    $finAffUnit .= substr($aff, 0, 5) . $separ . substr($aff, 5, 1) . $separ;
                    break;
                case "манни":
                case "менни":
                    $finAffUnit .= $aff[0] . $separ . substr($aff, 1, 3) . $separ . $aff[4] . $separ;
                    break;
                case "асси":
                case "есси":
                case "ясси":
                    $finAffUnit .= substr($aff, 0, 3) . $separ . $aff[3] . $separ;
                    break;
                case "ни":
                    $finAffUnit .= "н" . $separ . "и" . $separ;
                    break;
                case "малли":
                case "мелли":
                    $finAffUnit .= substr($aff, 0, 4) . $separ . "и" . $separ;
                    break;
                case "нӑҫем":
                case "нӗҫем":
                    $finAffUnit .= substr($aff, 0, 2) . $separ . substr($aff, 2, 3) . $separ;
                    break;
                case "пӑр":
                case "пӗр":
                    $finAffUnit .= $aff[0] . $separ . substr($aff, 1, 2) . $separ;
                    break;
                case "сене":
                    $finAffUnit .= "сем" . $separ . "е" . $separ;
                    break;
                case "аҫҫӗ":
                case "еҫҫӗ":
                    $finAffUnit .= substr($aff, 0, 2) . $separ . "ҫ" . $separ . "ӗ" . $separ;
                    break;
                case "мест":
                    $finAffUnit .= "мес" . $separ . "т" . $separ;
                    break;
                case "масть":
                    $finAffUnit .= "мас" . $separ . "ть" . $separ;
                    break;
                case "рӗ":
                    $finAffUnit .= $aff[0] . $separ . "ӗ" . $separ;
                    break;
                case "рӗҫ":
                case "чӗҫ":
                    $finAffUnit .= $aff[0] . $separ . "ӗ" . $separ . "ҫ" . $separ;
                    break;
                case "мӗ":
                    $finAffUnit .= "м" . $separ . "ӗ" . $separ;
                    break;
                case "ӗҫ":
                    $finAffUnit .= "ӗ" . $separ . "ҫ" . $separ;
                    break;
                case "атчӗ":
                case "етчӗ":
                    $finAffUnit .= $aff[0] . "тч" . $separ . "ӗ" . $separ;
                    break;
                case "чӗ":
                    if ($chast == "глагол" && $affunit[0] == "чӗ") {
                        $finAffUnit .= "ч" . $separ . "ӗ" . $separ;
                    } else {
                        $finAffUnit .= $aff . $separ;
                    }
                    break;
                default:
                    $finAffUnit .= $aff . $separ;
                    break;
            }
        }

        $finAffUnit = rtrim($finAffUnit, $separ); // Удаляем последний символ разделителя
        return $finAffUnit;
    }





}